Mark all CUDA ccalls as GC safe

vchuravy opened the pull request:
Using https://docs.julialang.org/en/v1/stdlib/Profile/#Triggered-During-Execution I observed an issue in #2261 where
we acquire a lock in `cuOccupancyMaxPotentialBlockSize` and seemingly get live locked. Other threads
suspend in waiting for GC and no forward progress is being made.

Manually marking `cuOccupancyMaxPotentialBlockSize` as safe to execute GC next to seemingly fixes the issue,
we could add a convenience macro for doing this.

---

Update by @maleadt: I made all ccalls GC safe, made sure callbacks are GC unsafe again.

x-ref https://github.com/JuliaLang/julia/pull/49933

vchuravy committed:
Mark cuOccupancyMaxPotentialBlockSize as GC threadsafe

maleadt committed:
Use a macro, and mark all libcuda calls gc-safe.

maleadt commented:
Pushed a macro version, and used that to mark all `libcuda` ccall's as GC-safe. I'm not confident this is allowed though, or at least not on all Julia versions, when doing a `ccall` that may call back into Julia.

Running the MWE from https://github.com/JuliaGPU/CUDA.jl/issues/2261 repeatedly on 1.8 doesn't seem to crash though.

maleadt committed:
Add explicit safepoints on 1.8.

vchuravy commented on lib/cudadrv/occupancy.jl:
```diff
@@ -36,7 +36,13 @@ end
+function _shmem_cint_cb(x::Cint)
+    # see @gcsafe_ccall documentation
+    @static if VERSION < v"1.9"
+        GC.safepoint()
```
Did you mean:

```
state = @ccall jl_gc_unsafe_enter()::Int8
```

and at the end:

```
@ccall jl_gc_unsafe_leave(state::Int8)
```

maleadt replied:
I didn't, but I guess I may be misunderstanding. I thought that acquiring a safepoint when GC is active would block / be sufficient of a workaround, also because you mentioned that manual action isn't needed on 1.10 now that we have safepoint_on_entry? i.e., there's no need to transition to GC unsafe here (we aren't invoking the runtime), only to block when the GC is active (which AFAIU you do by trying to acquire a safepoint).

vchuravy replied:
So the issue is that `cuOccupancyMaxPotentialBlockSize` is now executed with `GC safe` aka "GC allowed to run concurrenty"`. On 1.9+ upon re-entrancy into Julia through the cfunction thunk the Julia code is marked `GC unsafe` aka "GC must wait for safepoint to be reached before running". So everything should be fine. 
A thread is either blocked on the CUDA lock and marked `GC safe`, or a thread is within Julia code an can make forward-progress/hit a safepoint to execute GC.

On 1.8 `cfunction` doesn't handle the GC transition. So we are now marking a thread as `GC safe` and then switch back into Julia code while still being `GC safe`. If a user would insert an `@info` statement into the callback or do anything else interesting we may corrupt a concurrently running GC.

`GC.safepoint()` is part of `jl_gc_unsafe_enter`, but on it's own it is not sufficient. E.g. GC could start immediatly after the safepoint and we would have the same issue.

vchuravy replied:
I would use:
```
macro gcunsafe(expr)
    if VERSION < v"1.9.0"
        quote
            state = ccall(:jl_gc_unsafe_enter, Int8, ())
            val = $(esc(expr))
            ccall(:jl_gc_unsafe_leave, Cvoid, (Int8,), state)
            val
        end
    else
        return expr
    end
end
```

To wrap the callback functions. I can push that update?

maleadt replied:
> GC could start immediatly after the safepoint and we would have the same issue.

Ah right, that's the crucial bit I was missing. Thanks!

maleadt replied:
I'll push the update, as I'll have to rebase this branch on top of the header regenerate PR anyway.

vchuravy commented on lib/utils/call.jl:
```diff
@@ -165,6 +165,12 @@ render_arg(io, arg::Union{<:Base.RefValue, AbstractArray}) = summary(io, arg)
+# when callbacks occur, the code should ensure the GC is not running:
+# - on 1.10 and later, everything is fine because of safepoint_on_entry
+# - on 1.9, @cfunction-based callbacks are fine because they transition to gc_unsafe
+# - on 1.8 and earlier, the code should explicitly call GC.safepoint()!
```
On 1.9+ callbacks from C automatically transition the runtime to GC_unsafe
On 1.8 the code should explicitly transition to GC_unsafe

maleadt committed:
Format.

maleadt committed:
Update wrapper genarator to use gcsafe ccall macro.

maleadt committed:
Re-generate headers.

maleadt committed:
Avoid ambiguities due to double ccall.

maleadt committed:
Add entry safepoints to library callbacks.

maleadt committed:
Use a function decorator macro to transition back into GC unsafe domain.

maleadt committed:
Bugfix.

vchuravy commented on lib/utils/call.jl:
```diff
@@ -150,3 +160,102 @@ end
+        gc_state = @ccall(jl_gc_unsafe_enter()::Int8)
+        try
+            $body
+        finally
```
I was debating if the try/finally was necessary and decided that it wasn't. Since we can only catch exceptions in gc_unsafe. cc: @vtjnash for a double check on that.

vtjnash replied:
The `try` setup seems a bit risky to be present inside `jl_gc_safe_enter`, a `catch/finally` already implicitly reset the gc state upon leave (normal or error)

maleadt replied:
The `try` happens after `jl_gc_unsafe_enter`?

vtjnash replied:
Then I don't see how this is reachable, since cfunction already had to do an unsafe-enter to be able to dispatch into the runtime

vtjnash replied:
Oh, I see this is for v1.8. That seems okay. I am surprised that is still supported here

maleadt committed:
Fix subpackages.

Don't do the ccall change for cuQuantum as it doesn't support
the latest cuTENSOR, so it would be tricky maintaining both
in tree.

[skip julia]
[skip cuda]
